### DISCUSSION TITLE
D3 candlestick example error

### DISCUSSION THREAD

**Discussion by curtispaullithia** [d3/d3#3614] [Q&A] [answered]
Regarding using the candlestick example...  https://observablehq.com/@d3/candlestick-chart

Data looks like this:
{"Date":"2017-12-08T00:00:00.000Z","Open":170.490005,"High":171,"Low":168.820007,"Close":169.369995,"Adj Close":168.034302,"Volume":23355200}

Judging by error and code (and I may be incorrect) it seems like it's expecting Date to be a number instead of a string?

Do I need to fix the example code, or the example data or am I potentially using this thing incorrectly?

<img width="382" alt="image" src="https://user-images.githubusercontent.com/116024436/208363348-c1c5e766-c4f4-4685-911d-735b8dfbc5c2.png">

**Comment by curtispaullithia** [comment]
<img width="232" alt="image" src="https://user-images.githubusercontent.com/116024436/208364030-e56578f0-d061-4d54-bcfd-0db1ebd1262a.png">

**Comment by Fil** [comment, marked as answer]
In this notebook the cell that reads the data converts the date field from a string to a javascript Date, thanks to the typed: true option
~~~js
aapl = FileAttachment("aapl-2.csv").csv({typed: true}).then(data => data.slice(-120))
~~~

In a typical use of D3 you might use `d3.csv("url of the file", d3.autyType)` instead.

**Reply by curtispaullithia** [reply]
Is it typical that people load their data directly from a file?  I put the JSON example data/string into a var that I pass to the candle stick example function.  Maybe there is an issue with how the data is being read because the error appears to be caused by a value that is undefined and judging by the code it's related to the Date field in the example records.  I was wondering if somewhere the Date value needs to go through Date.parse() or something, maybe to turn it into a numeric value...but I think I need to get to the bottom of the undefined error.  Maybe it's simply not reading the input JSON data correctly or something like that.

**Reply by mbostock** [reply]
JSON doesn’t have a native representation of dates, so you’ll need to use the [JSON.parse _reviver_ option](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/parse) to turn date strings back into Date instances, or post-process your JSON (e.g., with array.map).

**Reply by curtispaullithia** [reply]
Nope the data is good and populating X.

**Reply by curtispaullithia** [reply]
Are you saying the example code doesn't work and I'll need to fix it to make it work?

**Reply by mbostock** [reply]
You haven’t posted your code, so I can’t say what you’ll need to change. But if you’re passing the CandlestickChart function data that has date strings instead of Date instances, it won’t work. And if you’re using JSON.parse or declaring your dates literally as strings, then you won’t have Date instances, so it won’t work. You should fix your data to have the proper types before you pass it to CandlestickChart.

**Reply by curtispaullithia** [reply]
Its not my data, it's the example data I downloaded from https://observablehq.com/@d3/candlestick-chart.  The resulting downloaded JSON file shows Date as a string.  Using unchanged example CandlestickChart function/code and simply passing the JSON text to it.  So the data is not correct?

**Reply by curtispaullithia** [reply]
<img width="500" alt="image" src="https://user-images.githubusercontent.com/116024436/208749191-ffce7ac1-abd0-4e8e-bcaf-8012f6c92ef3.png">

**Reply by curtispaullithia** [reply]
This is not the correct date format in the JSON data?  "Date":"2017-11-17T00:00:00.000Z"

**Reply by mbostock** [reply]
Not if you’re passing it directly to CandlestickChart, no. You have to pass e.g. `new Date("2017-11-17T00:00:00.000Z")` instead of `"2017-11-17T00:00:00.000Z"`. Or, like I said above, you need to use a reviver when you call JSON.parse, or otherwise convert your strings back to dates. JSON does not support dates, so you can’t use it as-is.

**Reply by curtispaullithia** [reply]
I see there was a D3.js update since this discussion.  I get what you are saying.
